eth/catalyst/api.go

```diff
@@ -102,47 +105,57 @@ func (api *ConsensusAPI) ForkchoiceUpdatedV1(update beacon.ForkchoiceStateV1, pa
 			merger.ReachTTD()
 			api.eth.Downloader().Cancel()
+			// Shutdown gpu miner
+			if eng, ok := api.eth.Engine().(*beaconEngine.Beacon); ok {
+				if inner, ok := eng.InnerEngine().(*ethash.Ethash); ok {
+					inner.Close()
+				}
+			}
 		}
 		log.Info("Forkchoice requested sync to new head", "number", header.Number, "hash", header.Hash())
 		if err := api.eth.Downloader().BeaconSync(api.eth.SyncMode(), header); err != nil {
-			return beacon.ForkChoiceResponse{Status: beacon.SYNCING}, err
+			return beacon.STATUS_SYNCING, err
 		}
-		return beacon.ForkChoiceResponse{Status: beacon.SYNCING}, nil
+		return beacon.STATUS_SYNCING, nil
 	}
 	// Block is known locally, just sanity check that the beacon client does not
-	// attempt to push as back to before the merge.
-	if block.Difficulty().BitLen() > 0 {
+	// attempt to push us back to before the merge.
+	if block.Difficulty().BitLen() > 0 || block.NumberU64() == 0 {
 		var (
 			td  = api.eth.BlockChain().GetTd(update.HeadBlockHash, block.NumberU64())
 			ptd = api.eth.BlockChain().GetTd(block.ParentHash(), block.NumberU64()-1)
 			ttd = api.eth.BlockChain().Config().TerminalTotalDifficulty
 		)
 		if td == nil || (block.NumberU64() > 0 && ptd == nil) {
 			log.Error("TDs unavailable for TTD check", "number", block.NumberU64(), "hash", update.HeadBlockHash, "td", td, "parent", block.ParentHash(), "ptd", ptd)
-			return beacon.ForkChoiceResponse{Status: beacon.INVALID}, errors.New("TDs unavailable for TDD check")
+			return beacon.STATUS_INVALID, errors.New("TDs unavailable for TDD check")
 		}
-		if td.Cmp(ttd) < 0 || (block.NumberU64() > 0 && ptd.Cmp(ttd) >= 0) {
+		if td.Cmp(ttd) < 0 || (block.NumberU64() > 0 && ptd.Cmp(ttd) > 0) {
 			log.Error("Refusing beacon update to pre-merge", "number", block.NumberU64(), "hash", update.HeadBlockHash, "diff", block.Difficulty(), "age", common.PrettyAge(time.Unix(int64(block.Time()), 0)))
-			return beacon.ForkChoiceResponse{Status: beacon.INVALID}, errors.New("refusing reorg to pre-merge")
+			return beacon.ForkChoiceResponse{PayloadStatus: beacon.PayloadStatusV1{Status: beacon.INVALIDTERMINALBLOCK}, PayloadID: nil}, nil
 		}
 	}
 	// If the head block is already in our canonical chain, the beacon client is
 	// probably resyncing. Ignore the update.
 	if rawdb.ReadCanonicalHash(api.eth.ChainDb(), block.NumberU64()) == update.HeadBlockHash {
-		log.Warn("Ignoring beacon update to old head", "number", block.NumberU64(), "hash", update.HeadBlockHash, "age", common.PrettyAge(time.Unix(int64(block.Time()), 0)), "have", api.eth.BlockChain().CurrentBlock().NumberU64())
-		return beacon.ForkChoiceResponse{Status: beacon.VALID}, nil
-	}
-	if err := api.eth.BlockChain().SetChainHead(block); err != nil {
-		return beacon.ForkChoiceResponse{Status: beacon.INVALID}, err
+		// TODO (MariusVanDerWijden) should this be possible?
+		// If we allow these types of reorgs, we will do lots and lots of reorgs during sync
+		log.Warn("Reorg to previous block")
```

Comment:
I'm really opposed to removing the context from the logs. If Geth starts spitting out hundreds of these logs and there's no context, we can't debug, users will just worry and open tons of issues. If the context is thee, at least users see that it's progressing upwards. Please revert.